Add specialized tablet_sstable_set

bhalevy opened the pull request:
Make a specialized sstable_set for tablets
via tablet_storage_group_manager::make_sstable_set.

This sstable set takes a snapshot of the storage_groups
(compound) sstable_sets and maps the selected tokens
directly into the tablet compound_sstable_set.

This sstable_set provides much more efficient access
to the table's sstable sets as it takes advantage of the disjointness
of sstable sets between tablets/storage_groups, and making it is cheaper
that rebuilding a complete partitioned_sstable_set from all sstables in the table.

Fixes #16876

Cassandra-stress setup:
```
$ sudo cpupower frequency-set -g userspace
$ build/release/scylla (developer-mode options) --smp=16 --memory=8G --experimental-features=consistent-topology-changes --experimental-features=tablets
cqlsh> CREATE KEYSPACE keyspace1 WITH replication={'class':'NetworkTopologyStrategy', 'replication_factor':1} AND tablets={'initial':2048};
$ ./tools/java/tools/bin/cassandra-stress write no-warmup n=10000000 -pop 'seq=1...10000000' -rate threads=128
$ scylla-api-client system drop_sstable_caches POST
$ ./tools/java/tools/bin/cassandra-stress read no-warmup duration=60s -pop 'dist=uniform(1..10000000)' -rate threads=128
$ scylla-api-client system drop_sstable_caches POST
$ ./tools/java/tools/bin/cassandra-stress mixed no-warmup duration=60s -pop 'dist=uniform(1..10000000)' -rate threads=128
```

Baseline (0a7854ea4d) vs. fix (0c2c00f01b)

Throughput (op/s):
workload | baseline | fix
---------|----------|----------
write | 76,806 | 100,787
read | 34,330 | 106,099
mixed | 32,195 | 79,246

bhalevy commented:
This PR supersedes https://github.com/scylladb/scylladb/pull/16918

avikivity commented on replica/table.cc:
```diff
@@ -2004,6 +2007,9 @@ void table::update_effective_replication_map(locator::effective_replication_map_
+
+        auto* tablet_sg_manager = dynamic_cast<tablet_storage_group_manager*>(_sg_manager.get());
+        tablet_sg_manager->update_effective_replication_map(_erm);
     }
```
Isn't handle_tablet_split_completion very similar? Should the functionality be merged?

bhalevy replied:
Yes, it can be done along with (or after) "table: move compaction_group_list and storage_group_vector to storage_group_manager"

bhalevy replied:
done in next version

avikivity commented on replica/table.cc:
```diff
@@ -565,34 +565,37 @@ class single_storage_group_manager final : public storage_group_manager {
+        : _t(t)
+        , _my_host_id(erm->get_token_metadata().get_my_id())
+        , _tablet_map(&erm->get_token_metadata().tablets().get_tablet_map(schema()->id()))
+    {}
```
Since you aren't storing erm, better to pass a regular reference.

Passing a pointer: the callee may use erm for a while, blocking changes
Passing a reference: the callee must use this atomically

This is similar to how a pointer can be null and a reference cannot.

Perhaps we should rename effective_replication_map_ptr to locked_effective_replication_map_ptr to make the consequences of holding it clearer.

bhalevy replied:
I think our conventions of _ptr naming are quite clear as they are, where holding a _ptr has a "capturing" effect.
There are exceptions to this rule like `shared_sstable` that maybe should have been called `sstable_ptr`, on the lines of `schema_ptr`, `token_metadata_ptr`, and `effective_replication_map_ptr`.

bhalevy replied:
I agree about the passing a reference in this case. will change

avikivity replied:
I routinely have to remind people to keep the ptr in scope, and to drop it when unneeded. It's clear, but it's clearly sending the wrong message. Holding on to a smart pointer keeps the value alive and doesn't impl any system-level effects. This one blocks the entire cluster from making forward progress on topology, and it's passed everywhere without a thought.

avikivity replied:
If there was a way to make identifiers in **bold**, this would be a good use case. People would think twice before passing a **`effective_replication_map_lock`**.

avikivity commented on replica/table.cc:
```diff
@@ -565,34 +565,37 @@ class single_storage_group_manager final : public storage_group_manager {
+
+    void update_effective_replication_map(const locator::effective_replication_map_ptr& erm) {
+        _tablet_map = &erm->get_token_metadata().tablets().get_tablet_map(schema()->id());
+    }
```
Similar here.

avikivity commented on replica/table.cc:
```diff
@@ -2004,6 +2007,9 @@ void table::update_effective_replication_map(locator::effective_replication_map_
         }
+
+        auto* tablet_sg_manager = dynamic_cast<tablet_storage_group_manager*>(_sg_manager.get());
+        tablet_sg_manager->update_effective_replication_map(_erm);
```
Let's make it a virtual function instead of this dynamic_cast.

bhalevy replied:
okay

avikivity commented on replica/compaction_group.hh:
```diff
@@ -218,11 +218,42 @@ public:
 
 class storage_group_manager {
+protected:
+    table& _t;
```
I think it belongs more in compaction than in database.

avikivity replied:
Although, maybe not. We abstracted tablets out of compaction with compaction::table_state. Now we're bringing it back. Let's limit the impact of tablets on modules.

We shouldn't just move things around for convenience, we have to think of modularity too.

bhalevy replied:
I can keep it in the table's compaction_group_manager implementations, but I thought that the classes here, being defined in replica/ header file, are in the replica database/table module, rather than the compaction module.  I agree they are borderline as they essentially define the border/interface between tables and compaction.

bhalevy replied:
I'll do that in the next version.
The tablet_sstable_set doesn't really need a table&, just a schema_ptr (for logging)

avikivity commented on replica/tablets.cc:
```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+        , _log2(sgm.log2_storage_groups())
+    {
+        _sstable_sets.reserve(sgm.storage_groups().size());
+        for (const auto& sg : sgm.storage_groups()) {
```
This isn't going to work well. Consider a 1 PB table, one table in the cluster. It will have 200,000 tablets. Of these only 100 will be on this shard. We'll iterate over 200,000 items to find 100.

avikivity replied:
The storage groups thing needs to maintain multiple data structures. A vector for fast lookup. A std::flat_map for in-order iteration.

raphaelsc replied:
Agreed, but it's orthogonal to this series since it's a preexisting problem. I am planning to fix it with this, see: https://github.com/raphaelsc/scylla/blob/ceaed7a842ec510db36b9d0119d010087da58565/utils/chunked_vector.hh#L502

avikivity replied:
ok

avikivity commented:
One thing we can do later, is have restricted versions of sstable_set that capture only a token range. Most sstable_sets will be happy to just capture a single tablet (repair, migration, BYPASS CACHE reads sent from storage_proxy). Only row_cache needs a full set.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6358/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6358/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7639/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6358/)
#### Build Details:
- Duration: 4 hr 2 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6402/)
#### Build Details:
- Duration: 12 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
In v2 ([15ad7e4](https://github.com/scylladb/scylladb/commit/15ad7e4c248c571b9a582626af28204899ff884d)):
- rebased
- tablet_storage_group_manager: precalculate my_host_id and _tablet_map
  -  pass e_r_m by ref (not using ref to e_r_m_ptr)
- table: move compaction_group_list and storage_group_vector to storage_group_manager
  - keep table& in derived storage_group_manager classes
  - move also `handle_tablet_split_completion` into `tablet_storage_group_manager`
- tablet_storage_group_manager: handle_tablet_split_completion: pre-calc new_tablet_count
  - cosmetic change only

avikivity commented on replica/compaction_group.hh:
```diff
@@ -219,15 +219,54 @@ public:
+protected:
+    // The compaction group list is only a helper for accessing the groups managed by the storage groups.
+    // The list entries are unlinked automatically when the storage group, they belong to, is removed.
+    mutable compaction_group_list _compaction_groups;
```
Why mutable?

raphaelsc replied:
I have a patch to get rid of this intrusive list. it was added for making integration easier, but it's dangerous as a compaction group alive but no longer in use could still be found in iteration. I don't remember why I had to make it mutable, but we'll remove it anyway.

raphaelsc replied:
we can rely only on a foreach_compaction_group or parallel_foreach_compaction_group for iterations.

raphaelsc replied:
@avikivity @bhalevy before I proceed with a patch (actually I have an ongoing one for it), I'd like to discuss this idea with you both. I am planning to rename storage_group to compaction_group_list, so we don't have to live with both terms (storage and compaction group). Think of compaction_group_list as our memtable_list. memtable_list may be composed of more than one memtable, all of which belong to the same replica table. Likewise, compaction_group_list is composed of more than one group, all of which belong to the same replica tablet. Makes sense, right? Then we can rename storage_group_manager to simply compaction_group_manager. Having both terms are very confusing, and making interfaces harder to reason about, as you don't know which one to work with.

raphaelsc replied:
So replica::table will have a set of compaction group lists, where each list belongs to an individual tablet replica.

raphaelsc replied:
storage_group_vector_ptr is replaced by compaction_group_lists = sparse_chunked_vector<compaction_group_list>; compaction_group_list = today's storage_group.

avikivity replied:
Some more ideas:

1. splittable_compaction_group, compaction_group

2. compaction_group base class, simple_compaction_group implementation, splitting_compaction_group implementation (with base, left, right simple_compaction_group members)

avikivity commented on replica/compaction_group.hh:
```diff
@@ -219,15 +219,54 @@ public:
+    // until the storage_group_manager finishes update_effective_replication_map
     virtual void update_effective_replication_map(const locator::effective_replication_map& erm) = 0;
-    virtual storage_group_vector make_storage_groups(compaction_group_list& list) const = 0;
+    virtual void make_storage_groups() = 0;
```
Move to constructor?

avikivity commented on replica/table.cc:
```diff
@@ -685,103 +736,105 @@ future<> storage_group::split(compaction_group_list& list, sstables::compaction_
 compaction_group& table::compaction_group_for_token(dht::token token) const noexcept {
     auto [idx, range_side] = storage_group_of(token);
-    auto& sg = *_storage_groups[idx];
+    auto& sg = *storage_groups()[idx];
```
Would be better not to expose storage_groups() at all, instead move the whole function to storage_group_manager. But not important for now.

avikivity commented on replica/table.cc:
```diff
@@ -1981,37 +2038,36 @@ void table::handle_tablet_split_completion(size_t old_tablet_count, const locato
+            return do_for_each(*groups, [] (std::unique_ptr<storage_group>& sg) {
+                return sg->main_compaction_group()->stop();
+            });
+        });
```
coroutine candidate for later.

bhalevy replied:
done in next version

avikivity commented on replica/compaction_group.hh:
```diff
@@ -217,6 +217,9 @@ public:
     future<> split(compaction_group_list&, sstables::compaction_type_options::split opt);
+
+    // Make an sstable set spanning all sstables in the storage_group
+    lw_shared_ptr<sstables::sstable_set> make_sstable_set() const;
```
Maybe return a const sstable_set? can be done later.

We should avoid making empty sstable sets and adding to them later, and try to have only sealed sets, where possible.

bhalevy replied:
At least the compound_sstable_set and tablet_sstable_sets that are returned on this path do not implement insert and erase so they are effectively immutable.

avikivity commented on replica/tablets.cc:
```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+        _sstable_sets.reserve(sgm.storage_groups().size());
+        for (const auto& sg : sgm.storage_groups()) {
+            if (sg) {
+                auto set = sg->make_sstable_set();
```
Wouldn't sg maintain an sstable set? do we have to create a new one?

Each time a tablet migrates we have to make_sstable_set() on all tablets, where we can just pick up an existing set.

bhalevy replied:
The name sstable_set is a bit misleading since the storage group returns a compound sstable set that just collects the underlying sstable sets using a shallow copy so it's relatively cheap.
That said, I think we can keep a pre-calculated lw_shared_ptr<sstable_set> per storage_group and just grab a reference to it, but we'll need to rebuild it when any of its underlying compaction_group's sstable_set:s is updated - this makes a tree-like refresh process during cache update where the leaf-level sstable sets are replaced and the compound sstable sets on top of them need to updated, up to the top level which is this tablet_sstable_set (in case tablets are used, otherwise it would be the single compaction_group compound_sstable_set which shortens this tree).

In any case, I suggest we save this as a followup since to keep this change relatively conservative and test the hell out of it while working on the next stage.

avikivity commented:
Please report on performance in the cover letter?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6411/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6411/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7660/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6411/)
#### Failed Tests (6253/22447):
- [test_count_and_group_by_row](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_count_and_group_by_row) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_count_and_group_by_row)
- [test_count_and_group_by_partition](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_count_and_group_by_partition) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_count_and_group_by_partition)
- [test_aggregation_inf_nan](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_aggregation_inf_nan) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_aggregation_inf_nan)
- [test_avg_decimal_2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_avg_decimal_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_avg_decimal_2)
- [test_sum_in_partition](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_sum_in_partition) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sum_in_partition)
- [test_sum_star](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_sum_star) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sum_star)
- [test_count_and_group_by_row](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_count_and_group_by_row_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_count_and_group_by_row)
- [test_count_and_group_by_partition](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_count_and_group_by_partition_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_count_and_group_by_partition)
- [test_aggregation_inf_nan](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_aggregation_inf_nan_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_aggregation_inf_nan)
- [test_avg_decimal_2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6411/testReport/junit/%28root%29/test_aggregate/Tests___Unit_Tests___test_avg_decimal_2_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_avg_decimal_2)
#### Build Details:
- Duration: 2 hr 31 min
- Builder: spider3.cloudius-systems.com

raphaelsc commented on replica/table.cc:
```diff
@@ -681,103 +740,119 @@ future<> storage_group::split(compaction_group_list& list, sstables::compaction_
+}
+
+future<> table::maybe_split_compaction_group_of(locator::tablet_id tablet_id) {
+    auto holder = async_gate().hold();
```
idea for later: hide compaction_group from external world.

bhalevy commented:
In v3 ([1ef8af9](https://github.com/scylladb/scylladb/commit/1ef8af95179ee9eb76b9967d752995ef252219ad)):
- rebased + retested
- table: coroutinize update_effective_replication_map
  - wait of deregistration of empty main compaction groups after split instead of deferring it to a background fiber
- table: move compaction_group_list and storage_group_vector to storage_group_manager
  - dropped `mutable` from _compaction_groups member
  - moved additional compaction-group related helpers into storage_group_manager to reduce the exposure of compaction_groups() even further.
- table: tablet_storage_group_manager: storage_group_of: do not validate in release build mode

raphaelsc commented on replica/database.hh:
```diff
@@ -449,11 +449,9 @@ private:
     compaction_manager& _compaction_manager;
     sstables::compaction_strategy _compaction_strategy;
+    // The storage_group_manager manages either a single storage_group for vnodes or per-table storage_group for tablets.
+    // It contains and manages both the compaction_groups list and the storage_groups vector.
```
per-tablet storage_group for tablets ?

raphaelsc commented on replica/tablets.cc:
```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+            return dht::maximum_token();
+        }
+        return dht::last_token_of_compaction_group(_log2, idx);
+    }
```
This works today, but it's assuming a particular implementation of tablet map. so knowledge duplication.

It seems you need a snapshot of tablet_map. Why don't you add tablet_map::token_whatever which stores only log2_tablets and provides all these token-related functions? tablet_map::token_whatever gets reused by tablet_map itself. Then you can call tablet_map::get_token_whatever() and store it here, which will represent the tablet map in a particular point in time.

bhalevy replied:
sounds like a good idea.
it may help convert to uneven tablet ranges if and when we get there

bhalevy replied:
that said, maybe we should do it as a follow-up since #16876 is P0 and I don't want to delay this series indefinitely (I'm sure it can be improved upon further and further)

bhalevy commented:
- rebased
- fixed comment
- fixed build error

raphaelsc commented:
The series look good to me. I am very tired today, so tomorrow morning I plan to review the incremental selection path.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6467/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6467/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6467/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7708/)
#### Build Details:
- Duration: 1 hr 58 min
- Builder: spider3.cloudius-systems.com

raphaelsc commented on replica/tablets.cc:
```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+
+            _cur_selector.emplace(_cur_set->make_incremental_selector());
+        }
+
```
we need a maybe_advance_selector() here, before we select below.

otherwise, we'll query a token that belongs to next group on the current one.

select() is called again when a next token is found that >= lowest_next_token, so we cannot advance selector after select() call below.

I think it's important that we add tests for such edge cases, when range scan is jumping across tablet boundaries for example. we want to make sure that data is properly returned.

e.g. 

read last token from tablet 0
read first token from tablet 1.

raphaelsc commented on replica/tablets.cc:
```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+        // jump to the next compaction_group sstable set.
+        dht::ring_position_ext lowest_next_position = res.next_position;
+        if (lowest_next_position.is_max()) {
+            _cur_set = {};
```
lowest_next_position.is_max() means we only switch selectors when last token is exhausted, is_max() refers to maximum token, not maximum token of current group.

also we need to do the check before selecting() as there's no guarantee last token of a group exists, and select() is only called after with a token of next group. and select() must be able to detect it and advance the selector before selecting.

bhalevy replied:
> lowest_next_position.is_max() means we only switch selectors when last token is exhausted, is_max() refers to maximum token, not maximum token of current group.

`lowest_next_position` is the `next_position` returned from the current selector.
It means that this selector is exhausted and we need to switch to the next one.
We return it when there are no more sstables matching the selected position, not the this position is last/maximum.
Or maybe I don't follow what you mean above.

raphaelsc replied:
I think "if (lowest_next_position.is_max()) {" will not work for switching selectors.

And you can't reset current selector after _cur_selector->select(pos), you have to do it before. 

You should check: does the current position span the current selector or should I move to the next one?

And only then call _cur_selector->select(pos).

You cannot predict which tokens will come next. next_position is only used as a hint for the reader to know when to invoke the selector again, after current position is >= next_position.

raphaelsc replied:
see this: https://github.com/scylladb/scylladb/pull/13666/commits/dc1c0708852646bc17e708bdad8dd7b96b2293f4#diff-970063d9f766ba24975f1955091d6bd0b071932333c47936fe2b4ff62032c580R2951

raphaelsc replied:
I understand now what you do :-) You're waiting for underlying selector to be exhausted, which is communicated with ring_position_view::max(). I think that could work too. I still think it would be good that we add a test for this, even though it's indirectly tested, edge cases may be hard to catch with tests that are not specialized. We don't have to do it now, but it would be good to track it in a github issue. We can easily even reuse existing tests that were written as bugs were found.

raphaelsc commented on replica/table.cc:
```diff
@@ -2002,7 +1997,7 @@ void table::update_effective_replication_map(locator::effective_replication_map_
                           _schema->ks_name(), _schema->cf_name(), old_tablet_count, new_tablet_count);
-            handle_tablet_split_completion(old_tablet_count, _erm->get_token_metadata().tablets().get_tablet_map(table_id));
+            co_await handle_tablet_split_completion(old_tablet_count, _erm->get_token_metadata().tablets().get_tablet_map(table_id));
         }
```
Ah found a problem with coroutinizing

1) tablet map is updated

2) handling of split completion yields, before new storage groups are set

3) token is mapped to tablet id which storage group doesn't exist yet

what we can do is to defer tablet map update to after handling of split completion

we can rely on table::update_effective_replication_map keeping old erm alive till completion.

bhalevy replied:
okay, can do that

raphaelsc commented on replica/table.cc:
```diff
@@ -1962,6 +1962,8 @@ void table::handle_tablet_split_completion(size_t old_tablet_count, const locato
         }
+        // Remove old main groups, they're unused, but they need to be deregistered properly
+        co_await sg->main_compaction_group()->stop();
         unsigned first_new_id = id << growth_factor;
```
we cannot yield here, the remapping must be atomic. ideally, we don't futurize handle_tablet_split_completion, so we make it explicit.

bhalevy replied:
We can keep a future<> for the background work.
The stop() calls should return a ready future anyhow (in non-debug mode), no?
since the main_compaction_group is known to be empty.

xemul commented:
For future references:
| branch | sct test id | test |
|------------|------------------------------------------------------|--------------------------------------------|
| master | 8d7cb3d2-1f22-408c-8b8e-9453a566d1cc | test_latency_write_with_nemesis |
| master | 87185741-7cb0-4f4a-9656-4d0d7cc084e0 | test_latency_read_with_nemesis |
| master | ce5d62ee-2ed6-4db3-af39-dec050286151 | test_latency_mixed_with_nemesis |
| this pr  | b6153834-6ceb-44a5-adc3-2be3e318f94d | test_latency_write_with_nemesis |
| this pr  | 099b1ad4-1c22-46af-be26-66b8df6e2b91 | test_latency_read_with_nemesis |
| this pr  | b265335c-9aa1-4507-8d40-083525d77702 | test_latency_mixed_with_nemesis |

bhalevy commented:
in v5 ([4bba763](https://github.com/scylladb/scylladb/commit/4bba7636cb1d478f3dbbd52efea8e25a8b0bb99f)):
- rebased
- table: coroutinize update_effective_replication_map
  - collect and return unready futures from `compaction_group.stop()` in `handle_tablet_split_completion` so not to yield
  - awaited in update_effective_replication_map
- table: tablet_storage_group_manager: make tablet_sstable_set
  - instantiate a tablet_map in tablet_sstable_set to map token <-> tablet_id back and forth
  - tablet_incremental_select: invalidate _cur_selector if given `pos` if out of range for it
  - handle pos == dht::maximum case
  - rename `lowest_next_position` variable to `next_position` and improved comments around it

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6578/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6578/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7823/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6578/)
#### Failed Tests (490/13061):
- [test_allow_filtering_indexed_no_filtering](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_indexed_no_filtering) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_indexed_no_filtering)
- [test_allow_filtering_indexed_filtering_required](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_indexed_filtering_required) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_indexed_filtering_required)
- [test_allow_filtering_indexed_a_and_k](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_indexed_a_and_k) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_indexed_a_and_k)
- [test_allow_filtering_multi_column](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_multi_column) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_multi_column)
- [test_allow_filtering_indexed_a_and_c](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_indexed_a_and_c) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_indexed_a_and_c)
- [test_allow_filtering_indexed_s_and_c](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_indexed_s_and_c) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_indexed_s_and_c)
- [test_allow_filtering_indexed_two_indexes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_indexed_two_indexes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_indexed_two_indexes)
- [test_allow_filtering_prefix](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_prefix) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_prefix)
- [test_allow_filtering_pk_in](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_pk_in) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_pk_in)
- [test_allow_filtering_indexed_no_filtering](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6578/testReport/junit/%28root%29/test_allow_filtering/Tests___Unit_Tests___test_allow_filtering_indexed_no_filtering_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_allow_filtering_indexed_no_filtering)
#### Build Details:
- Duration: 14 hr
- Builder: spider6.cloudius-systems.com

raphaelsc commented on replica/tablets.cc:
```diff
@@ -308,4 +310,281 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+            // Set the next token to point to the possible storage group.
+            // It will be considered later on when the _cur_set is exhausted
+            _lowest_next_token = find_lowest_next_token(_tset.first_token_of(idx+1));
+        }
```
what if we're on the last tablet? can it happen idx+1 results in out of bound access? maybe find_lowest_next_token() should receive idx, rather than token. and if we're on last tablet (_tmap.last_tablet() == idx), then max token is returned.

bhalevy replied:
It's prepared for that:
```c++
        if (idx >= _sstable_sets.size()) {
            return dht::maximum_token();
        }
```

raphaelsc replied:
where is that? I cannot find it.

raphaelsc replied:
maybe it's uncommitted?

bhalevy replied:
I used to have logic for that but when I change the selector to call the tablet_map `first_token_of` this protection is now gone.
I'll fix in the next version.

raphaelsc commented:
just had one comment, other than that, looks good to me.

bhalevy commented:
CI failures, e.g.
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6578/artifact/testlog/x86_64/debug/alternator.run.3.log
```
Scylla version 5.5.0~dev-0.20240211.bcfd0238d3ad with build-id b09765cf1edd2abde2301f205f5be244e0a5f188 starting ...

dht/token.hh:227:44: runtime error: signed integer overflow: 9223372036854775807 + 1 cannot be represented in type 'int64_t' (aka 'long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior dht/token.hh:227:44 in 
Aborting on shard 1.
Backtrace:
  0xd3130ca
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x48b902c
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x48b8c83
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x46f2a98
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x4723785
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481ea29
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481ed6e
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481eb7e
  /lib64/libc.so.6+0x3dbaf
  /lib64/libc.so.6+0x8e883
  /lib64/libc.so.6+0x3dafd
  /lib64/libc.so.6+0x2687e
  0xd38ceea
  0xd38ac94
  0xd39f363
  0xd3a06ef
  0xd3a04df
  0x122fedf4
  0xe8163b8
  0xf65d959
  0xf6d76a1
  0xf05d0e7
  0xf05f631
  0xf065dc4
  0xf682e74
  0x11df58da
  0xd5565d7
  0xd5562b0
  0xd55609d
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x47155f8
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x471dbb8
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x4721c50
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x4821726
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481f8d0
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481f7c0
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481ef2c
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x44c63e8
  /lib64/libc.so.6+0x8c946
  /lib64/libc.so.6+0x11285f
```
Decoded:
```
[Backtrace #0]
__sanitizer::Abort() at ??:?
__sanitizer::Die() at ??:?
__ubsan::ScopedReport::~ScopedReport() at ??:?
void handleIntegerOverflowImpl<__ubsan::Value>(__ubsan::OverflowData*, unsigned long, char const*, __ubsan::Value, __ubsan::ReportOptions) at ??:?
__ubsan_handle_add_overflow at ??:?
dht::next_token(dht::token const&) at ././dht/token.hh:227
 (inlined by) locator::tablet_map::get_first_token(locator::tablet_id) const at ./locator/tablets.cc:203
replica::tablet_sstable_set::first_token_of(unsigned long) const at ./replica/tablets.cc:393
 (inlined by) replica::tablet_incremental_selector::select(dht::ring_position_view const&) at ./replica/tablets.cc:508
sstables::sstable_set::incremental_selector::select(dht::ring_position_view const&) const at ./sstables/sstable_set.cc:211
sstables::incremental_reader_selector::create_new_readers(std::optional<dht::ring_position_view> const&) at ./sstables/sstable_set.cc:817
mutation_reader_merger::maybe_add_readers(std::optional<dht::ring_position_view> const&) at ./readers/combined.cc:312
mutation_reader_merger at ./readers/combined.cc:470
make_combined_reader(seastar::lw_shared_ptr<schema const>, reader_permit, std::unique_ptr<reader_selector, std::default_delete<reader_selector> >, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>) at ./readers/combined.cc:687
sstables::sstable_set::make_local_shard_sstable_reader(seastar::lw_shared_ptr<schema const>, reader_permit, nonwrapping_interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>, sstables::read_monitor_generator&, seastar::noncopyable_function<bool (sstables::sstable const&)> const&) const at ./sstables/sstable_set.cc:1381
operator() at ./db/view/view.cc:1932
```

raphaelsc commented:
> CI failures, e.g. https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6578/artifact/testlog/x86_64/debug/alternator.run.3.log
> 
> ```
> Scylla version 5.5.0~dev-0.20240211.bcfd0238d3ad with build-id b09765cf1edd2abde2301f205f5be244e0a5f188 starting ...
> 
> dht/token.hh:227:44: runtime error: signed integer overflow: 9223372036854775807 + 1 cannot be represented in type 'int64_t' (aka 'long')
> SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior dht/token.hh:227:44 in 
> Aborting on shard 1.
> Backtrace:
>   0xd3130ca
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x48b902c
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x48b8c83
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x46f2a98
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x4723785
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481ea29
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481ed6e
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481eb7e
>   /lib64/libc.so.6+0x3dbaf
>   /lib64/libc.so.6+0x8e883
>   /lib64/libc.so.6+0x3dafd
>   /lib64/libc.so.6+0x2687e
>   0xd38ceea
>   0xd38ac94
>   0xd39f363
>   0xd3a06ef
>   0xd3a04df
>   0x122fedf4
>   0xe8163b8
>   0xf65d959
>   0xf6d76a1
>   0xf05d0e7
>   0xf05f631
>   0xf065dc4
>   0xf682e74
>   0x11df58da
>   0xd5565d7
>   0xd5562b0
>   0xd55609d
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x47155f8
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x471dbb8
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x4721c50
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x4821726
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481f8d0
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481f7c0
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x481ef2c
>   /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x44c63e8
>   /lib64/libc.so.6+0x8c946
>   /lib64/libc.so.6+0x11285f
> ```
> 
> Decoded:
> 
> ```
> [Backtrace #0]
> __sanitizer::Abort() at ??:?
> __sanitizer::Die() at ??:?
> __ubsan::ScopedReport::~ScopedReport() at ??:?
> void handleIntegerOverflowImpl<__ubsan::Value>(__ubsan::OverflowData*, unsigned long, char const*, __ubsan::Value, __ubsan::ReportOptions) at ??:?
> __ubsan_handle_add_overflow at ??:?
> dht::next_token(dht::token const&) at ././dht/token.hh:227
>  (inlined by) locator::tablet_map::get_first_token(locator::tablet_id) const at ./locator/tablets.cc:203
> replica::tablet_sstable_set::first_token_of(unsigned long) const at ./replica/tablets.cc:393
>  (inlined by) replica::tablet_incremental_selector::select(dht::ring_position_view const&) at ./replica/tablets.cc:508
> sstables::sstable_set::incremental_selector::select(dht::ring_position_view const&) const at ./sstables/sstable_set.cc:211
> sstables::incremental_reader_selector::create_new_readers(std::optional<dht::ring_position_view> const&) at ./sstables/sstable_set.cc:817
> mutation_reader_merger::maybe_add_readers(std::optional<dht::ring_position_view> const&) at ./readers/combined.cc:312
> mutation_reader_merger at ./readers/combined.cc:470
> make_combined_reader(seastar::lw_shared_ptr<schema const>, reader_permit, std::unique_ptr<reader_selector, std::default_delete<reader_selector> >, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>) at ./readers/combined.cc:687
> sstables::sstable_set::make_local_shard_sstable_reader(seastar::lw_shared_ptr<schema const>, reader_permit, nonwrapping_interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>, sstables::read_monitor_generator&, seastar::noncopyable_function<bool (sstables::sstable const&)> const&) const at ./sstables/sstable_set.cc:1381
> operator() at ./db/view/view.cc:1932
> ```

I think it happens here:

`        return dht::next_token(get_last_token(tablet_id(size_t(id) - 1)));`

when accessing out of bound.

we must have check_tablet_id(id); in get_first_token().

bhalevy commented:
> when accessing out of bound.
> 
> we must have check_tablet_id(id); in get_first_token().

I'm fixing that in the next version based on your suggestion to pass idx to `find_lowest_next_token` and let it return maximum_token if idx is out of bounds.
I added non-release mode checks in first_token_of and last_token_of to abort if they were called with out-of-bound idx.

raphaelsc commented:
> > when accessing out of bound.
> > we must have check_tablet_id(id); in get_first_token().
> 
> I'm fixing that in the next version based on your suggestion to pass idx to `find_lowest_next_token` and let it return maximum_token if idx is out of bounds. I added non-release mode checks in first_token_of and last_token_of to abort if they were called with out-of-bound idx.

great, thanks

bhalevy commented:
In v6 ([777bd6a](https://github.com/scylladb/scylladb/commit/777bd6a5f79300b14a1313b4f449a518955bddb9)):
- rebased
- pass idx to find_lowest_next_token and let it return maximum_token if idx is out of bounds.
- added non-release mode checks in first_token_of and last_token_of to abort if they were called with out-of-bound idx.
- added initialization of cumulative _size and _bytes_on_disk members to 0.

raphaelsc reviewed (approved):
looks good

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6670/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6670/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7901/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6670/)
#### Build Details:
- Duration: 2 hr 6 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
ping @avikivity please re-review and merge if everything is fine

avikivity commented:
There is still no performance report

bhalevy commented:
> There is still no performance report

I thought https://github.com/scylladb/scylladb/pull/17149#issuecomment-1935830051 is sufficient.

avikivity commented:
> > There is still no performance report
> 
> I thought [#17149 (comment)](https://github.com/scylladb/scylladb/pull/17149#issuecomment-1935830051) is sufficient.

1. Lot's of data but no analysis, what should I conclude from it? What problem is the patch solving?
2. Nothing in the cover letter.

xemul commented:
Ran perf-simple-query with #17473 against master and this PR with 1M partitions, BYPASS CACHE and different number partitions in memtable -- the new option that made the population phase create more than 1 sstable (auto compaction was disabled as well). Test cases are

* vnodes -- master, without tablets
* tablets-1 -- master, with tablets and 1 initial tablet
* tablets-100 -- master, with tablets and 100 initial tablets
* tablets-100-patch -- this PR , with tablets and 100 initial tablets

By and large, the options were `-c8 -m8G --partitions 10000000 --enable-cache=0 --bypass-cache --flush-partitions N --flush`

With no flushes in the middle (just 1 flush after population)
![image](https://github.com/scylladb/scylladb/assets/4498177/db6f4409-889d-4144-81b0-e094477e91e1)

With flush every 50k partitions (200 flushes while populating)
![image](https://github.com/scylladb/scylladb/assets/4498177/1347bee7-2250-42be-bf8c-3c644d9227d3)

With flush every 5k partitions (2000 flushes while populating)
![image](https://github.com/scylladb/scylladb/assets/4498177/356251cf-711c-47be-ae01-96cf54bb502f)

avikivity commented:
What's the y axis?

bhalevy commented:
> What's the y axis?

transactions per second (I asked @xemul the same question)

bhalevy commented:
@avikivity ping

avikivity commented:
Why?

bhalevy commented:
Cc @Deexie re: https://github.com/scylladb/scylladb/issues/16877

Deexie commented on replica/compaction_group.hh:
```diff
@@ -222,12 +222,51 @@ public:
 using storage_group_vector = utils::chunked_vector<std::unique_ptr<storage_group>>;
```
Please note that this commit introduces construction loop:

In `table` constructor we create 
```
, _sg_manager(make_storage_group_manager())
```

In `make_storage_group_manager` we create `tablet_storage_group_manager/single_storage_group_manager`.
In their constructors we create `compaction_group` which in turn calls `compaction_manager::add` in its constructor:
```
void compaction_manager::add(table_state& t) {
    auto [_, inserted] = _compaction_state.try_emplace(&t, t);
    if (!inserted) {
        on_internal_error(cmlog, format("compaction_state for table {} [{}] already exists", t, fmt::ptr(&t)));
    }
}
```
If table_state isn't inserted properly we hit:
https://github.com/scylladb/scylladb/blob/e3de30f94392672fa858df39e42735eb7b58662e/compaction/table_state.hh#L62-L73

which uses
```
table_state::get_group_id()  ->
table::compaction_groups() ->
storage_group_manager::compaction_groups()
```

And `storage_group_manager` is being created at this point so it's nullptr.

bhalevy replied:
thanks.  will fix in next version.

bhalevy commented:
In v7 ([0c2c00f](https://github.com/scylladb/scylladb/commit/0c2c00f01b52357e94dfed0c4c549f3fb6701087))
- rebased
- compaction_group::table_state: get_group_id: become self-sufficient

bhalevy commented:
Latest benchmark results with cassandra-stress (will also summarize in this PR's description):

Before (0a7854ea4d)
```
$ sudo cpupower frequency-set -g userspace
$ scylla-developer-mode --smp=16 --memory=8G --experimental-features=consistent-topology-changes --experimental-features=tablets
cqlsh> CREATE KEYSPACE keyspace1 WITH replication={'class':'NetworkTopologyStrategy', 'replication_factor':1} AND tablets={'initial':2048};
$ ./tools/java/tools/bin/cassandra-stress write no-warmup n=10000000 -pop 'seq=1...10000000' -rate threads=128
Op rate                   :   76,806 op/s  [WRITE: 76,806 op/s]
Partition rate            :   76,806 pk/s  [WRITE: 76,806 pk/s]
Row rate                  :   76,806 row/s [WRITE: 76,806 row/s]
Latency mean              :    1.6 ms [WRITE: 1.6 ms]
Latency median            :    1.7 ms [WRITE: 1.7 ms]
Latency 95th percentile   :    3.0 ms [WRITE: 3.0 ms]
Latency 99th percentile   :    4.1 ms [WRITE: 4.1 ms]
Latency 99.9th percentile :    5.5 ms [WRITE: 5.5 ms]
Latency max               :   20.0 ms [WRITE: 20.0 ms]
Total partitions          : 10,000,000 [WRITE: 10,000,000]
Total errors              :          0 [WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 00:02:10

$ scylla-api-client system drop_sstable_caches POST
$ ./tools/java/tools/bin/cassandra-stress read no-warmup duration=60s -pop 'dist=uniform(1..10000000)' -rate threads=128
Partition rate            :   34,330 pk/s  [READ: 34,330 pk/s]
Row rate                  :   34,330 row/s [READ: 34,330 row/s]
Latency mean              :    3.7 ms [READ: 3.7 ms]
Latency median            :    3.5 ms [READ: 3.5 ms]
Latency 95th percentile   :    7.0 ms [READ: 7.0 ms]
Latency 99th percentile   :    9.1 ms [READ: 9.1 ms]
Latency 99.9th percentile :   12.1 ms [READ: 12.1 ms]
Latency max               :   34.7 ms [READ: 34.7 ms]
Total partitions          :  2,070,374 [READ: 2,070,374]
Total errors              :          0 [READ: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 00:01:00

$ scylla-api-client system drop_sstable_caches POST
$ ./tools/java/tools/bin/cassandra-stress mixed no-warmup duration=60s -pop 'dist=uniform(1..10000000)' -rate threads=128
Op rate                   :   32,195 op/s  [READ: 16,066 op/s, WRITE: 16,128 op/s]
Partition rate            :   32,195 pk/s  [READ: 16,066 pk/s, WRITE: 16,128 pk/s]
Row rate                  :   32,195 row/s [READ: 16,066 row/s, WRITE: 16,128 row/s]
Latency mean              :    3.9 ms [READ: 5.2 ms, WRITE: 2.7 ms]
Latency median            :    3.5 ms [READ: 4.8 ms, WRITE: 2.7 ms]
Latency 95th percentile   :    8.5 ms [READ: 10.1 ms, WRITE: 4.9 ms]
Latency 99th percentile   :   12.1 ms [READ: 13.7 ms, WRITE: 6.1 ms]
Latency 99.9th percentile :   17.4 ms [READ: 19.2 ms, WRITE: 7.9 ms]
Latency max               :   36.3 ms [READ: 36.3 ms, WRITE: 22.6 ms]
Total partitions          :  1,947,575 [READ: 971,911, WRITE: 975,664]
Total errors              :          0 [READ: 0, WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 00:01:00
```

After (0c2c00f01b)
```
$ sudo cpupower frequency-set -g userspace
$ scylla-developer-mode --smp=16 --memory=8G --experimental-features=consistent-topology-changes --experimental-features=tablets
cqlsh> CREATE KEYSPACE keyspace1 WITH replication={'class':'NetworkTopologyStrategy', 'replication_factor':1} AND tablets={'initial':2048};
$ ./tools/java/tools/bin/cassandra-stress write no-warmup n=10000000 -pop 'seq=1...10000000' -rate threads=128
Op rate                   :  100,787 op/s  [WRITE: 100,787 op/s]
Partition rate            :  100,787 pk/s  [WRITE: 100,787 pk/s]
Row rate                  :  100,787 row/s [WRITE: 100,787 row/s]
Latency mean              :    1.3 ms [WRITE: 1.3 ms]
Latency median            :    1.1 ms [WRITE: 1.1 ms]
Latency 95th percentile   :    2.6 ms [WRITE: 2.6 ms]
Latency 99th percentile   :    3.1 ms [WRITE: 3.1 ms]
Latency 99.9th percentile :    4.4 ms [WRITE: 4.4 ms]
Latency max               :   25.6 ms [WRITE: 25.6 ms]
Total partitions          : 10,000,000 [WRITE: 10,000,000]
Total errors              :          0 [WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 00:01:39

$ scylla-api-client system drop_sstable_caches POST
$ ./tools/java/tools/bin/cassandra-stress read no-warmup duration=60s -pop 'dist=uniform(1..10000000)' -rate threads=128
Op rate                   :  106,099 op/s  [READ: 106,099 op/s]
Partition rate            :  106,099 pk/s  [READ: 106,099 pk/s]
Row rate                  :  106,099 row/s [READ: 106,099 row/s]
Latency mean              :    1.2 ms [READ: 1.2 ms]
Latency median            :    1.1 ms [READ: 1.1 ms]
Latency 95th percentile   :    2.3 ms [READ: 2.3 ms]
Latency 99th percentile   :    3.1 ms [READ: 3.1 ms]
Latency 99.9th percentile :    4.7 ms [READ: 4.7 ms]
Latency max               :   68.2 ms [READ: 68.2 ms]
Total partitions          :  6,374,957 [READ: 6,374,957]
Total errors              :          0 [READ: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 00:01:00

$ scylla-api-client system drop_sstable_caches POST
$ ./tools/java/tools/bin/cassandra-stress mixed no-warmup duration=60s -pop 'dist=uniform(1..10000000)' -rate threads=128
Op rate                   :   79,246 op/s  [READ: 39,648 op/s, WRITE: 39,598 op/s]
Partition rate            :   79,246 pk/s  [READ: 39,648 pk/s, WRITE: 39,598 pk/s]
Row rate                  :   79,246 row/s [READ: 39,648 row/s, WRITE: 39,598 row/s]
Latency mean              :    1.6 ms [READ: 1.8 ms, WRITE: 1.3 ms]
Latency median            :    1.6 ms [READ: 1.9 ms, WRITE: 1.2 ms]
Latency 95th percentile   :    3.3 ms [READ: 3.5 ms, WRITE: 2.5 ms]
Latency 99th percentile   :    4.0 ms [READ: 4.2 ms, WRITE: 3.0 ms]
Latency 99.9th percentile :    5.5 ms [READ: 6.0 ms, WRITE: 4.3 ms]
Latency max               :   42.5 ms [READ: 42.5 ms, WRITE: 39.9 ms]
Total partitions          :  4,819,553 [READ: 2,411,286, WRITE: 2,408,267]
Total errors              :          0 [READ: 0, WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 00:01:00
```

avikivity commented:
Please drop the latency stats from the cover letter, latency is meaningless for a throughput test (latency is derived from Little's Law).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7168/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7168/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8347/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/309/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7168/)
#### Build Details:
- Duration: 3 hr 57 min
- Builder: i-0bda2e31d16e801a6 (m5ad.12xlarge)

bhalevy commented:
@scylladb/scylla-maint please consider merging

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -2878,7 +2878,7 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
             for (auto it = table_erms.begin(); it != table_erms.end(); ) {
                 auto& cf = db.find_column_family(it->first);
-                cf.update_effective_replication_map(std::move(it->second));
+                co_await cf.update_effective_replication_map(std::move(it->second));
```
This entire function `storage_service::replicate_to_all_cores` tries very hard to make updates atomic/isolated i.e. that every shard observes all cf/erm changes at a single instant. It goes a long way to construct a `pending_table_erms` struct and so on, so it can then move those erms to the observable state in a single task without a yield.

And then this simple `co_await` abandons all of that work. Breaking the purpose of the multi-phase approach that this function took.

So now we can end up in a state, for example, when we have a table with materialized view attached to it, where the base erm reflects a topology change, while view erm still doesn't -- or vice versa. And data path queries will observe such inconsistent state.
